_posts/2024-10-31-isa-roadmap.markdown

```diff
@@ -0,0 +1,140 @@
+author: "David Chisnall"
+---
+
+The CHERIoT ISA has been fairly stable for a while now.
+The main addition has been the newer sentry types that allow us to differentiate between function pointers and return addresses and to better protect interrupt-disabled functions from abuse.
+There have also been a few small fixes to corner cases.
+You can read these in the change log in the [current live version of the specification](https://cheriot.org/cheriot-sail/cheriot-architecture.pdf) (Appendix A).
+
+We aim to stabilise a version 1.0 and have a project [tracking open issues to be resolved before then](https://github.com/orgs/CHERIoT-Platform/projects/1).
+
+The 1.0 release will include a stable abstract machine and something that we expect to support in the toolchain for a long time.
+This will be referred to as the Xcheriotv1 extension to RISC-V.
+
+CHERIoT does not compete with the RISC-V Zcheri standards
+---------------------------------------------------------
+
+People often ask me if they should use CHERIoT or the standard RISC-V CHERI32 extension.
+The answer, of course, is 'yes'.
+
+CHERIoT is a mature extension that various people have been running software on for several years and that has been refined based on that experience, but it's also an extension that is tailored for small devices.
+We made a number of choices that let us scale down to 3-stage single-issue pipelines with tens of KiBs of RAM, which would be the wrong approach on multicore, multi-GHz superscalar out-of-order architectures with TiBs of RAM.
+
+The CHERI standardisation effort, in contrast, has to produce something that works everywhere.
+It is defining a set of extensions under the Zcheri umbrella.
+It also has to produce something that is a minimum viable product, with a focus on memory safety.
+A lot of things in CHERIoT have demonstrated value in our use case will not make it into a 1.0 release of the standard CHERI extensions.
+
+The things added to RISC-V by CHERIoT can be split into several categories:
+
+ - Things that we think everyone should do.
+ - Things where we've made compromises in a direction are necessary for small devices.
+ - Things where we've co-designed the hardware and software and, while we think they're the right approach in a purist sense, may not be the most compatible options.
+
+We've been working with the RISC-V International CHERI SIG and TG on ensuring that the Zcheripurecap extension is a subset of CHERIoT.
+We believe that, in the current drafts, everything in our first category is in the RVI standard and everything in the second and third categories are permitted.
+
+This will become important for CHERIoT 2.0.
+
+CHERIoT ISA v 2.0
+-----------------
+
+Our goal with the 2.0 release of the CHERIoT ISA is to rebase on top of the official Zcheripurecap extension, once this has been ratified.
+
+We expect to be able to provide compatibility between 1.0 and 2.0, such that anything that isn't machine code can run on both.
```

Comment:
```suggestion
We expect to be able to provide source compatibility between 1.0 and 2.0, such that anything that isn't machine code can run on both.
```

Reply:
I didn't want to say 'source' because assembly isn't usually in the space of 'source'.

Reply:
That's fine; stet.